io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -829,8 +839,8 @@ reschedule !simstate@SimState{ threads, timers, curTime = time } =
         !_ <- mapM_ (\(SomeTVar tvar) -> unblockAllThreadsFromTVar tvar) written
 
             -- Check all fired threadDelays
-        let wakeupThreadDelay = [ tid | TimerThreadDelay tid <- fired ]
-            wakeup            = wakeupThreadDelay ++ wakeupSTM
+        let wakeupThreadDelay = [ (tid, tmid) | TimerThreadDelay tid tmid <- fired ]
+            wakeup            = fst `fmap` wakeupThreadDelay ++ wakeupSTM
```

Comment:
We can also filter here all the `threadDelays` that belong to threads that still exist. Would that be simpler than the filter when exception logic?

Reply:
I think it makes more sense to do that earlier, when we deliver the exception. At the end its the exception that terminates the thread delay.

Reply:
Done.